internal/check/binop.go

```diff
@@ -0,0 +1,80 @@
+package check
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/ory/keto/internal/check/checkgroup"
+	"github.com/ory/keto/internal/relationtuple"
+	"github.com/ory/keto/ketoapi"
+)
+
+type binaryOperator = func(ctx context.Context, checks []checkgroup.CheckFunc) checkgroup.Result
+
+func or(ctx context.Context, checks []checkgroup.CheckFunc) checkgroup.Result {
+	if len(checks) == 0 {
+		return checkgroup.ResultNotMember
+	}
+
+	resultCh := make(chan checkgroup.Result, len(checks))
+	childCtx, cancelFn := context.WithCancel(ctx)
+	defer cancelFn()
+
+	for _, check := range checks {
```

Comment:
Spawning a potential unlimited amount of go routines that execute very little logic or which cause a lot of load should be avoided at all cost.

Calling things asynchronous makes sense if the workload is worth the overhead. If we are spawning a go routine to return a (more or less static) value, the overhead of spawning a goroutine is not worth it.

When using concurrency to execute costly functions - usually functions that use RPC calls such as SQL, redis, protobuf, ... - we need to limit how much load a single node can generate.

Please re-evaluate the concurrency pattern here and decide whether it is needed or not. If it is needed, please replace this with a configurable worker pool. If you need inspiration: https://github.com/aeneasr/pool

Reply:
Good point. We have implemented it this way because at a later point we want to off-load some of the concurrent processing to other Keto nodes, essentially a cluster mode.

Reply:
The parallelism of the binary operations is directly bound by the branching of the AST. So if we put an upper bound on the number of expressions e.g., 'or' together, we automatically limit this here.

I am talking about expressions like this:
```ts
this.related.P.traverse(p => p.permits.view(ctx)) ||
this.related.A.includes(ctx.subject) ||
this.related.B.includes(ctx.subject) ||
this.related.C.includes(ctx.subject) ||
this.related.D.includes(ctx.subject)
```
which will result in all 5 checks executing in parallel.

I still think that executing these in parallel is warranted, since each check will ultimately execute one or more database queries.

I think the better approach is to limit the branching in the AST (and leave this code as is). Right now, only the depth of the AST is limited.

WDYT?

Reply:
While that might limit the concurrency of a singular check, it won't limit the concurrency of millions of checks running in a multi-tenant system. If enough requests hit, and parallelism is not bound by a worker pool, we will run into DoS issues.

I thus recommend implementing a worker pool regardless of the decisions in the AST. Limiting depth and branching is a good idea in general as an unbound model could also lead to DoS attacks against Ory Cloud (by using a very wide policy or potentially recursive policy)

Reply:
OK makes sense. This will then be a global worker pool for all checks, right? Not scoped to one top-level check request?

Reply:
@aeneasr I implemented a worker pool (basically https://gobyexample.com/worker-pools), PTAL :smiley: